tubemq-connectors/tubemq-connector-flink/src/main/java/org/apache/flink/connectors/tubemq/TubemqSinkFunction.java

```diff
@@ -0,0 +1,178 @@
+
+    private static final String TID = "tid";
+
+    /**
+     * The address of tubemq master, format eg: 127.0.0.1:8080,127.0.0.2:8081.
+     */
+    private final String masterAddress;
+
+    /**
+     * The topic name.
+     */
+    private final String topic;
+
+    /**
+     * The tid of this topic
+     */
+    private final String tid;
+    /**
+     * The serializer for the records sent to pulsar.
+     */
+    private final SerializationSchema<T> serializationSchema;
+
+    /**
+     * The tubemq producer.
+     */
+    private transient MessageProducer producer;
+
+    /**
+     * The tubemq session factory.
+     */
+    private transient MessageSessionFactory sessionFactory;
+
+    /**
+     * The maximum number of retries.
+     */
+    private final int maxRetries;
+
+    public TubemqSinkFunction(String topic,
+                              String masterAddress,
+                              SerializationSchema<T> serializationSchema,
+                              Configuration configuration) {
+        Preconditions.checkNotNull(topic,
+            "The topic must not be null.");
+        Preconditions.checkNotNull(masterAddress,
+            "The master address must not be null.");
+        Preconditions.checkNotNull(serializationSchema,
+            "The serialization schema must not be null.");
+        Preconditions.checkNotNull(configuration,
+            "The configuration must not be null.");
+
+        this.topic = topic;
+        this.masterAddress = masterAddress;
+        this.serializationSchema = serializationSchema;
+        this.tid = configuration.getString(TubemqOptions.TID);
+        this.maxRetries = configuration.getInteger(MAX_RETRIES);
+    }
+
+    @Override
+    public void snapshotState(FunctionSnapshotContext functionSnapshotContext) {
+        // Nothing to do.
+    }
+
+    @Override
+    public void initializeState(FunctionInitializationContext functionInitializationContext) {
+        // Nothing to do.
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+
+        TubeClientConfig tubeClientConfig = new TubeClientConfig(masterAddress);
+        this.sessionFactory = new TubeSingleSessionFactory(tubeClientConfig);
+        this.producer = sessionFactory.createProducer();
+        HashSet<String> hashSet = new HashSet<>();
+        hashSet.add(topic);
+        producer.publish(hashSet);
+    }
+
+    @Override
+    public void invoke(T in, Context context) throws Exception {
+
+        int retries = 0;
+        Exception exception = null;
+
+        while (maxRetries <= 0 || retries < maxRetries) {
+
+            try {
+                byte[] body = serializationSchema.serialize(in);
+                Message message = new Message(topic, body);
+                if (StringUtils.isNotBlank(tid)) {
+                    message.setAttrKeyVal(TID, tid);
```

Comment:
If the message carries tid, and the business needs to filter consume according to this field, it is also necessary to call putSystemHeader(), so that the consumer can do filter consumption

Reply:
ok